gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/domain_service/subscription/RejectSubscriptionDomainServiceImpl.java

```diff
@@ -0,0 +1,204 @@
+        this.auditDomainService = auditDomainService;
+        this.userCrudService = userCrudService;
+    }
+
+    @Override
+    public SubscriptionEntity rejectSubscription(ExecutionContext executionContext, String subscriptionId, AuditActor auditActor) {
+        log.debug("Close subscription {}", subscriptionId);
+        Optional<Subscription> optSubscription;
+        try {
+            optSubscription = subscriptionRepository.findById(subscriptionId);
+        } catch (TechnicalException e) {
+            log.error("An error occurs while trying to get subscription [subscriptionId={}]", subscriptionId, e);
+            throw new TechnicalManagementException(e);
+        }
+
+        final SubscriptionEntity subscriptionEntity = optSubscription
+            .map(SubscriptionAdapter.INSTANCE::toEntity)
+            .orElseThrow(() -> new SubscriptionNotFoundException(subscriptionId));
+
+        return rejectSubscription(executionContext, subscriptionEntity, auditActor);
+    }
+
+    @Override
+    public SubscriptionEntity rejectSubscription(
+        ExecutionContext executionContext,
+        SubscriptionEntity subscriptionEntity,
+        AuditActor auditActor
+    ) {
+        if (subscriptionEntity == null) {
+            // TODO: find a better exception
+            throw new IllegalArgumentException("Subscription should not be null");
+        }
+        checkSubscriptionStatus(subscriptionEntity);
+        checkPlanStatus(subscriptionEntity);
+        var rejectedSubscriptionEntity = subscriptionEntity.rejectBy(auditActor.userId());
+        var subscription = SubscriptionAdapter.INSTANCE.fromEntity(rejectedSubscriptionEntity);
+
+        try {
+            subscriptionRepository.update(subscription);
+
+            triggerNotifications(executionContext, rejectedSubscriptionEntity);
+            createAudit(executionContext, subscriptionEntity, rejectedSubscriptionEntity, auditActor);
+        } catch (TechnicalException e) {
+            log.error("An error occurs while trying to save subscription [subscriptionId={}]", subscription.getId(), e);
+            throw new TechnicalManagementException(e);
+        }
+
+        return rejectedSubscriptionEntity;
+    }
+
+    private void triggerNotifications(ExecutionContext executionContext, SubscriptionEntity rejectedSubscriptionEntity) {
+        var apiHook = new SubscriptionRejectedApiHookContext(
+            rejectedSubscriptionEntity.getApiId(),
+            rejectedSubscriptionEntity.getApplicationId(),
+            rejectedSubscriptionEntity.getPlanId()
+        );
+        triggerNotificationDomainService.triggerApiNotification(executionContext, apiHook);
+        var applicationHook = new SubscriptionRejectedApplicationHookContext(
+            rejectedSubscriptionEntity.getApplicationId(),
+            rejectedSubscriptionEntity.getApiId(),
+            rejectedSubscriptionEntity.getPlanId()
```

Comment:
By reading it again, I find it very weird that applicationId and apiId are not in the same order for apiNotification or applicationNotification.
But not to do in this PR as I think it would be best to move all notification related logic into the notification domain service

Reply:
to be honest, my test was failing because I did not put them in the right order...
Semanticly, the first param is the one associated to the notification, but yes it's weird when using

Reply:
I did the same mistake too (reversing the param order). Which means other devs will do too 😅